crates/task-impls/src/quorum_vote/handlers.rs

```diff
@@ -135,6 +135,52 @@ async fn handle_quorum_proposal_validated_drb_calculation_seed<
     Ok(())
 }
 
+/// Handles calling add_epoch_root and sync_l1 on Membership if necessary.
+async fn handle_quorum_proposal_validated_add_epoch_root<
+    TYPES: NodeType,
+    I: NodeImplementation<TYPES>,
+    V: Versions,
+>(
+    proposal: &QuorumProposal2<TYPES>,
+    task_state: &mut QuorumVoteTaskState<TYPES, I, V>,
+    leaf_views: &[LeafInfo<TYPES>],
+) {
+    let decided_block_number = leaf_views
+        .last()
+        .unwrap()
+        .leaf
+        .block_header()
+        .block_number();
+
+    // Skip if this is not the expected block.
+    if task_state.epoch_height != 0 && (decided_block_number + 3) % task_state.epoch_height == 0 {
+        let next_epoch_number = TYPES::Epoch::new(
+            epoch_from_block_number(decided_block_number, task_state.epoch_height) + 1,
+        );
+
+        let membership_reader = task_state.membership.read().await;
```

Comment:
Possibly personal preference, but I think it is more idiomatic to prefer introducing a scoped block over explicit `drop`s. Something like:

```rust
let write_callback = {
    let membership_reader = task_state.membership.read().await;
    membership_reader.add_epoch_root(next_epoch_number, proposal.block_header.clone())
}

```